generator/src/lib.rs

```diff
@@ -38,33 +42,58 @@ impl quote::ToTokens for CType {
 
 named!(ctype<&str, CType>,
     alt!(
-        tag!("ULL") => { |_| CType::U64 } |
-        tag!("U") => { |_| CType::U32  }
+        value!(CType::U64, complete!(tag!("ULL"))) |
+        value!(CType::U32, complete!(tag!("U")))
     )
 );
+
 named!(cexpr<&str, (CType, String)>,
-       alt!(
-           map!(cfloat, |f| (CType::Float, format!("{:.2}", f))) |
-           inverse_number
-       )
+    alt!(
+        map!(cfloat, |f| (CType::Float, format!("{:.2}", f))) |
+        inverse_number |
+        typed_cnumber
+    )
+);
+
+named!(typed_cnumber<&str, (CType, String)>,
+    do_parse!(
+        num: digit1 >>
+        typ: ctype >>
+        ((typ, num.to_string()))
+    )
+);
+
+named!(hexadecimal_number<&str, (CType, Literal)>,
+    map!(
+        preceded!(
+            alt!(tag!("0x") | tag!("0X")),
+            pair!(hex_digit1, ctype)
+        ),
+        |(num, typ)| (typ, {
+            // This is how lit::SynInt reconstructs an integer
+            // because proc_macros2::Literal lacks a way to construct
+            // hexadecimal literals.
+            let stream = format!("0x{}", num).parse::<TokenStream>().unwrap();
+            match stream.into_iter().next().unwrap() {
+                TokenTree::Literal(l) => l,
+                _ => unreachable!(),
+            }
```

Comment:
We do the same hack in `cexpr`, IMO we should just combine this all and return a `Literal` ro `TokenStream` directly from the macro.